client/my-sites/checkout/composite-checkout/wpcom/types/wpcom-store-state.ts

```diff
@@ -1,3 +1,210 @@
+): ManagedContactDetailsShape< B > {
+	const tldExtraFields: ManagedContactDetailsTldExtraFieldsShape< B > = {};
+
+	if ( y.tldExtraFields?.ca ) {
+		if ( x.tldExtraFields?.ca ) {
+			tldExtraFields.ca = {
+				lang: f( x.tldExtraFields.ca.lang, y.tldExtraFields.ca.lang ),
+				legalType: f( x.tldExtraFields.ca.legalType, y.tldExtraFields.ca.legalType ),
+				ciraAgreementAccepted: f(
+					x.tldExtraFields.ca.ciraAgreementAccepted,
+					y.tldExtraFields.ca.ciraAgreementAccepted
+				),
+			};
+		} else {
+			tldExtraFields.ca = y.tldExtraFields.ca;
+		}
+	} else if ( x.tldExtraFields?.ca ) {
+		tldExtraFields.ca = {
+			lang: g( x.tldExtraFields.ca.lang ),
+			legalType: g( x.tldExtraFields.ca.legalType ),
+			ciraAgreementAccepted: g( x.tldExtraFields.ca.ciraAgreementAccepted ),
+		};
+	}
+
+	if ( y.tldExtraFields?.uk ) {
+		if ( x.tldExtraFields?.uk ) {
+			tldExtraFields.uk = {
+				registrantType: f( x.tldExtraFields.uk.registrantType, y.tldExtraFields.uk.registrantType ),
+				registrationNumber: f(
+					x.tldExtraFields.uk.registrationNumber,
+					y.tldExtraFields.uk.registrationNumber
+				),
+				tradingName: f( x.tldExtraFields.uk.tradingName, y.tldExtraFields.uk.tradingName ),
+			};
+		} else {
+			tldExtraFields.uk = y.tldExtraFields.uk;
+		}
+	} else if ( x.tldExtraFields?.uk ) {
+		tldExtraFields.uk = {
+			registrantType: g( x.tldExtraFields.uk.registrantType ),
+			registrationNumber: g( x.tldExtraFields.uk.registrationNumber ),
+			tradingName: g( x.tldExtraFields.uk.tradingName ),
+		};
+	}
+
+	if ( y.tldExtraFields?.fr ) {
+		if ( x.tldExtraFields?.fr ) {
+			tldExtraFields.fr = {
+				registrantType: f( x.tldExtraFields.fr.registrantType, y.tldExtraFields.fr.registrantType ),
+				trademarkNumber: f(
+					x.tldExtraFields.fr.trademarkNumber,
+					y.tldExtraFields.fr.trademarkNumber
+				),
+				sirenSirat: f( x.tldExtraFields.fr.sirenSirat, y.tldExtraFields.fr.sirenSirat ),
+			};
+		} else {
+			tldExtraFields.fr = y.tldExtraFields.fr;
+		}
+	} else if ( x.tldExtraFields?.fr ) {
+		tldExtraFields.fr = {
+			registrantType: g( x.tldExtraFields.fr.registrantType ),
+			trademarkNumber: g( x.tldExtraFields.fr.trademarkNumber ),
+			sirenSirat: g( x.tldExtraFields.fr.sirenSirat ),
+		};
+	}
+
+	return {
+		firstName: f( x.firstName, y.firstName ),
+		lastName: f( x.lastName, y.lastName ),
+		organization: f( x.organization, y.organization ),
+		email: f( x.email, y.email ),
+		alternateEmail: f( x.alternateEmail, y.alternateEmail ),
+		phone: f( x.phone, y.phone ),
+		phoneNumberCountry: f( x.phoneNumberCountry, y.phoneNumberCountry ),
+		address1: f( x.address1, y.address1 ),
+		address2: f( x.address2, y.address2 ),
+		city: f( x.city, y.city ),
+		state: f( x.state, y.state ),
+		postalCode: f( x.postalCode, y.postalCode ),
+		countryCode: f( x.countryCode, y.countryCode ),
+		fax: f( x.fax, y.fax ),
+		vatId: f( x.vatId, y.vatId ),
+		tldExtraFields,
+	};
+}
+
+function flattenManagedContactDetailsShape< A, B >(
```

Comment:
This might also benefit from a unit test? I mean, it's pretty simple logic, but maybe it would be easy enough to add some because of that?

Reply:
Writing this test exposed an outright bug in the implementation. :)